Cannot use type defined later as function return type

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn foo() -> S {
    S { a: 15 }
}

struct S {
    a: i32,
}

```

I expected to see this happen: No errors

Instead, this happened:
```rust
test.rs:13:16: error: failed to lookup HIR 20 for node 'S'
   13 | fn bidule() -> S {
      |                ^
test.rs:13:1: error: failed to resolve return type
   13 | fn bidule() -> S {
      | ^
```
This might be due to the fact that the return type gets resolved in the top-level resolver and not the item's one.

--- a/gcc/testsuite/rust/compile/issue-1361.rs
+++ b/gcc/testsuite/rust/compile/issue-1361.rs
@@ -1,0 +1,8 @@
+// { dg-options "-w" }
+fn foo() -> S {
+    S { a: 15 }
+}
+
+struct S {
+    a: i32,
+}

